CRM: Add webpack build step

gogdzl opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

In the CRM Gulp is used to build assets (minify JS, compile SASS into CSS). 
This PR aims to change this from Gulp to webpack.
We are doing this because there is an effort to remove gulp from the Monorepo (p9dueE-4RS-p2).

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Adds webpack build step to the CRM
* Removes gulp 
* Hides built files via .gitignore and defines which files to include in production in .gitattributes.
* Includes some changes to existing CSS and min files to make compilation and minification smoother.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
p1672845316058569-slack-C04J4UG3FLG

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Spin up a version of this branch
* Run `jetpack build plugins/crm` for a development build, and `jetpack build --production plugins/crm` to test the build as if it were in a production environment. 
* For each JS file inside a `js` folder, a minified JS file named `[original filename].min.js` should be created (in the same folder) (actual minification only happens if running a production build).
* For each SCSS file inside a `sass` folder, a minified CSS file named `[original filename].min.css` should be created (in a folder names `css` in the same level as the `sass` folder) (again actual minification only happens if running a production build).
* Additional license and CSS map files (and rtl files) will also be created depending on whether it's a production or development build.
* If testing the builds in succession, files should be cleaned up as part of the 'clean' script (except after a production build but a development build will take care of that).
* Finally, you should be able to navigate the CRM dashboard using the built version of CRM from this branch (for example using JN and the beta plugin) with no issues.


Other things of note:
* The functions `zeroBSCRMJS_listView_generic_editdirectlink` and `zeroBSCRMJS_listView_invoice_bulkActionTitle_delete` in `ZeroBSCRM.admin.listview.js` were duplicated which broke the build (one of each removed but worth a check to see the content in the remaining one is correct as `zeroBSCRMJS_listView_generic_editdirectlink` appeared to be copy pastes from different functions).
* Also one CSS fix was needed too in `zbs-exitform.css` to add a missing opening bracket.
* There are several removed min.css (and a min.js) file where they are minified via script anyway. A few files have been unminified where only a minified version existed, to aid in the minification process via script. In the process a couple of CSS issues were resolved as well.

gogdzl committed:
Remove gulp dependencies

gogdzl committed:
Add basic webpack build step

gogdzl committed:
changelog

gogdzl committed:
Add basic build steps

coder-karen committed:
Merge remote-tracking branch 'origin/trunk' into crm/add/webpack_build_step

coder-karen committed:
Moving to use jetpackWebpackConfig for the webpack build

anomiex commented on projects/plugins/crm/webpack.config.js:
```diff
@@ -0,0 +1,135 @@
+	node: false,
+	plugins: [
+		...jetpackWebpackConfig.StandardPlugins( {
+			DependencyExtractionPlugin: false,
```
It would probably be a good idea to look into using this at some point. Done right, it gets rid of the need for stuff like lines 86 and 87 below.

coder-karen replied:
Yes I'd like to come back to this, possibly after the move to Webpack if the current implementation isn't causing issues.

anomiex commented on projects/plugins/crm/webpack.config.js:
```diff
@@ -0,0 +1,135 @@
+							loader: 'sass-loader',
+							options: {
+								sassOptions: {
+									outputStyle: 'compressed',
```
This shouldn't be needed. Normally we allow sass to output uncompressed CSS, and rely on Terser (included in `jetpackWebpackConfig.optimization` above) to do the minification in production mode.

coder-karen committed:
Fixing a build issue related to semantic-ui, and a missing character in CSS file issue

coder-karen committed:
Adding in 'concurrently' package

anomiex commented on projects/plugins/crm/package.json:
```diff
@@ -1,29 +1,36 @@
+		"build": "pnpm run clean && pnpm run build-client",
+		"build-client": "NODE_PATH=\"$PWD/node_modules\" webpack",
+		"build-concurrently": "pnpm run clean && concurrently 'pnpm:build-client' 'pnpm:build-php'",
+		"build-production-concurrently": "pnpm run clean && concurrently 'NODE_ENV=production BABEL_ENV=production pnpm run build-client'",
```
If you want you could probably drop the use of `concurrently` here. You only have the one build command.

anomiex commented on projects/plugins/crm/webpack.config.js:
```diff
@@ -0,0 +1,129 @@
+		entry: getSassEntries(),
+		output: {
+			...crmWebpackConfig.output,
+			assetModuleFilename: '[path]/../css/[name].min.css',
```
I don't think this does what you intend it to do. `assetModuleFilename` is for stuff like images included from the JS.

Your best bet for fixing the css directories is to change the return value from  `getSassEntries()`. Right now it looks like
```js
{
    './sass/foo': './sass/foo.scss',
    './modules/foo/sass/bar': './modules/foo/sass/bar.scss',
}
```
If you change them to be like this, that should do what you want
```js
{
    './css/foo': './sass/foo.scss',
    './modules/foo/css/bar': './modules/foo/sass/bar.scss',
}
```

coder-karen replied:
Brilliant, thank you.

coder-karen committed:
Making sure CSS files load in correct directory

coder-karen commented on projects/plugins/crm/package.json:
```diff
@@ -1,29 +1,33 @@
-		"clean": "npx wp-env clean all"
+		"build": "pnpm run clean && pnpm run build-client",
+		"build-client": "NODE_PATH=\"$PWD/node_modules\" webpack",
+		"clean": "rm -rf build/",
```
I suspect I don't need this clean script either as there is no build directory.

anomiex replied:
True. Or you could adjust it to remove the files that actually are built.

coder-karen committed:
Changing composer build command to reflect what is included

anomiex commented on projects/plugins/crm/composer.json:
```diff
@@ -16,7 +16,7 @@
 		],
 		"build-production": [
-			"pnpm run build-production-concurrently"
+			"pnpm run build"
```
As this is now, your production build won't actually set the `NODE_ENV=production` and `BABEL_ENV=production` environment variables for webpack to know to do a production build. You could prefix the command here, or you could make a "build-production" script in package.json, either way works.

coder-karen replied:
Gotcha.

coder-karen committed:
Adding the welcome-to-zbs folder to the list of what is minified, plus fixing a CSS issue so it works

coder-karen committed:
Fixing scripts and removing duplicated functions to allow the build to run in production

coder-karen committed:
Updating webpack.config.js to add min to the minified CSS, and updating the gitignore file

coder-karen committed:
Cleaning up added comments in zbs-exitform.css

coder-karen committed:
Fixing merge conflict in pnpm-lock.yaml

coder-karen committed:
Cleaning up the clean command and ignoring more build generated files

coder-karen committed:
Removing several min files that get built via script, and adding css versions of some min versions

anomiex commented on projects/plugins/crm/package.json:
```diff
@@ -1,29 +1,34 @@
+		"@babel/core": "7.20.7",
+		"@babel/preset-env": "7.20.2",
+		"@babel/register": "7.18.9",
+		"@babel/runtime": "7.20.7",
```
Let's pick up the changes from #28682 here
```suggestion
		"@babel/core": "7.20.12",
		"@babel/preset-env": "7.20.2",
		"@babel/register": "7.18.9",
		"@babel/runtime": "7.20.13",
```

anomiex commented on projects/plugins/crm/package.json:
```diff
@@ -1,29 +1,34 @@
+		"build": "pnpm run clean && pnpm run build-client",
+		"build-client": "NODE_PATH=\"$PWD/node_modules\" webpack",
+		"build-production": "NODE_ENV=production BABEL_ENV=production pnpm run build",
+		"clean": "rm -rf **/css/**/*.rtl.* .cache/ **/css/*.min.css **/js/*.min.js.* **/js/*.min.js  **/js/**/*.min.js.map **/css/**/*.css.map",
```
The `**` here probably doesn't work right. Bash by default doesn't have globstar enabled.

This seems to work
```suggestion
		"clean": "rm -rf css/*.min.css* css/*.min.rtl.css css/integrations/*.min.css* css/integrations/*.min.rtl.css css/welcome-to-zbs/*.min.css* css/welcome-to-zbs/*.min.rtl.css js/*.min.js* js/welcome-to-zbs/wizard2.min.js*",
```
It looks like there are still some `.min.css` files checked in or we could do like
```suggestion
		"clean": "find css js modules \( -name '*.min.css*' -o -name '*.min.rtl.css' -o -name '*.min.js*' \) -delete"
```

coder-karen replied:
> It looks like there are still some .min.css files checked in or we could do like

Yes that would be a good 'next' project, to do the same for all folders as was done with welcome-to-zbs, depending on what the CRM preferences are.

> The ** here probably doesn't work right. Bash by default doesn't have globstar enabled.

Ah! Must have been a setting at my end that helped it work. Thank you. I'll commit a new change, as it looks even better to just delete the whole /integrations folder (only created when the CSS is being compiled and minified now).

anomiex reviewed (commented):
One other thing, to keep the minified files in the mirror repo you'll want to add entries in `projects/plugins/crm/.gitattributes` setting `production-include` for them all.

coder-karen committed:
Adding entries in .gitattributes

coder-karen committed:
[not verified] Update projects/plugins/crm/package.json

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

coder-karen committed:
Updating package.json with a cleaner clean script

coder-karen committed:
[not verified] Updating pnpm-lock.yaml

coder-karen committed:
Adding production-includes for three vendor packages and including license.txt files

Adding production-includes for two vendor packages.

[not verified] Production-including the dompdf package in this PR to aid in testing

Adding a production-include for a dmopdf dependency

A .gitattributes amendment - adding license.txt and two more vendor packages to production-include

anomiex commented on projects/plugins/crm/.gitattributes:
```diff
@@ -19,13 +19,22 @@ node_modules         export-ignore
 vendor/composer/**                              production-include
 vendor/jetpack-autoloader/**                    production-include
+vendor/dompdf/**                                production-include
+vendor/phenx/**                                production-include
```
I note that dompdf has two more deps (masterminds/html5 and sabberworm/php-css-parser) that aren't being included yet. They should probably be added too, unless we're sure that no CRM code path will ever use the dompdf features that need them.

coder-karen replied:
At this stage I'm not sure about that so it makes sense to add them in (added now).

anomiex commented on projects/plugins/crm/.gitattributes:
```diff
@@ -35,15 +44,19 @@ vendor/jetpack-autoloader/**                    production-include
+**/css/**/*.min.css.map                         production-exclude
+**/js/*.min.js.map                              production-exclude
+**/js/**/*.min.js.map                           production-exclude
+**/js/*.min.js.LICENSE.txt                      production-exclude
```
You might want to include these LICENSE.txt files in the distributed plugin, as licenses on bundled JS packages may require that the notices be kept with the distribution.

OTOH, I think currently the only things being included in any of these are `/*!` comments from stuff in this repo. If you change those comments to remove the `!`, the files might go away entirely.

coder-karen replied:
Ah good to know. That could be something we can follow up with (changing comments), possibly even next week, but adding the license.txt line to production-include to catch it anyway.

coder-karen committed:
Adding options to allow some function names to be exported and therefore defined

anomiex commented on projects/plugins/crm/js/ZeroBSCRM.admin.global.js:
```diff
@@ -2771,3 +2771,19 @@ const jpcrm = {
+		zbsJS_updateScreenOptions, zeroBSCRMJS_buildScreenOptionsTableColumns, zeroBSCRMJS_buildScreenOptionsGenerics, zeroBSCRMJS_rest_retrieveCompany, 
+		zeroBSCRMJS_bindGlobalObjFuncs, zeroBSCRMJS_bindGlobalContactFuncs, zeroBSCRMJS_obj_viewLink, zeroBSCRMJS_obj_editLink, zeroBSCRMJS_bindGlobalDismiss, 
+		zbscrm_JS_bindCloseLogs, jpcrm_bind_licensing_modals, jpcrm_bind_customfield_csv_builders, jpcrm_bind_generic_window_opening, zbscrm_JS_isCalypso, 
+		jpcrm_set_jpcrm_transient };
```
Is it really necessary to export all of these? Any that are only called from within the same bundle shouldn't need exporting.

coder-karen replied:
Noting we talked about this here - p1675685735295849-slack-C04J4UG3FLG . Largely just me misunderstanding some terminology (new commit pushed now).

coder-karen committed:
Removing unnecessary functions from being exported

gogdzl commented:
I tested the whole system and it looks like it works. The files are being generated and placed in the correct places. It feels like the CRM is almost working: there are many errors throughout, but it is really hard to discern what could be caused by the move to webpack from what could be caused by the move to the Monorepo. With that in mind, I say that we can move forward and fix any outstanding problems caused by webpack in later PRs if we get any.

Not a blocker, but I notice that we are missing the file `jetpack-crm/css/font-awesome.min.css`. Originally this file was 'hard-coded' into our css directory. I don't see any problems using that same approach and just adding the file to the correct place. We can do this in this PR or in a later one.

Since I opened this PR, I cannot add a review and approve it, but consider it approved by me.

coder-karen commented:
>  With that in mind, I say that we can move forward and fix any outstanding problems caused by webpack in later PRs if we get any.

Yes, sounds good.

> Not a blocker, but I notice that we are missing the file jetpack-crm/css/font-awesome.min.css. Originally this file was 'hard-coded' into our css directory. I don't see any problems using that same approach and just adding the file to the correct place. We can do this in this PR or in a later one.

Yes let's follow up separately now - adding it back in the current clean script will delete it on every build - it seems better to either add the unminified version and let that get minified for the builds, or link to the CDN instead of adding it in the source code? We can discuss further on another PR though, I'll add one shortly.

> Since I opened this PR, I cannot add a review and approve it, but consider it approved by me.

Thanks @gogdzl !

coder-karen committed:
Changing version numbers to 5.5.4

Changing version numbers to 5.5.4

Changing version numbers to 5.5.4

Changing next release version number

Changing version in composer.json as well.

Correcting versions in changelog / readme

Changing the significance of this PR to patch
